storage/badger/execution_data_tracker.go

```diff
@@ -0,0 +1,446 @@
+// This method first inserts a record of the blob at the specified block height.
+// It then checks if the current block height is greater than the previously recorded
+// latest height for this CID. If so, it updates the latest height to the current block height.
+//
+// Parameters:
+// - tx: The BadgerDB transaction in which the operation is performed.
+// - blockHeight: The height at which the blob was observed.
+// - c: The CID of the blob to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlob(tx *badger.Txn, blockHeight uint64, c cid.Cid) error {
+	err := operation.InsertBlob(blockHeight, c)(tx)
+	if err != nil {
+		return fmt.Errorf("failed to add blob record: %w", err)
+	}
+
+	var latestHeight uint64
+	err = operation.RetrieveTrackerLatestHeight(c, &latestHeight)(tx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("failed to get latest height: %w", err)
+		}
+	} else {
+		// don't update the latest height if there is already a higher block height containing this blob
+		if latestHeight >= blockHeight {
+			return nil
+		}
+	}
+
+	err = operation.UpsertTrackerLatestHeight(c, blockHeight)(tx)
+	if err != nil {
+		return fmt.Errorf("failed to set latest height value: %w", err)
+	}
+
+	return nil
+}
+
+// trackBlobs tracks multiple blob CIDs at the specified block height.
+//
+// Parameters:
+// - blockHeight: The height at which the blobs were observed.
+// - cids: The CIDs of the blobs to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlobs(blockHeight uint64, cids ...cid.Cid) error {
+	cidsPerBatch := storage.CidsPerBatch
+	maxCidsPerBatch := getBatchItemCountLimit(s.db, 2, storage.BlobRecordKeyLength+storage.LatestHeightKeyLength+8)
+	if maxCidsPerBatch < cidsPerBatch {
+		cidsPerBatch = maxCidsPerBatch
+	}
+
+	for len(cids) > 0 {
+		batchSize := cidsPerBatch
+		if len(cids) < batchSize {
+			batchSize = len(cids)
+		}
+		batch := cids[:batchSize]
+
+		if err := operation.RetryOnConflict(s.db.Update, func(txn *badger.Txn) error {
+			for _, c := range batch {
+				if err := s.trackBlob(txn, blockHeight, c); err != nil {
+					return fmt.Errorf("failed to track blob %s: %w", c.String(), err)
+				}
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		cids = cids[batchSize:]
+	}
+
+	return nil
+}
+
+// batchDelete deletes multiple blobs from the storage in a batch operation.
+//
+// Parameters:
+// - deleteInfos: Information about the blobs to be deleted, including their heights and whether to delete the latest height record.
+//
+// Ensures that all specified blobs are deleted from the storage.
+//
+// Returns an error if the batch deletion fails.
+func (s *ExecutionDataTracker) batchDelete(deleteInfos []*storage.DeleteInfo) error {
+	for _, dInfo := range deleteInfos {
+		err := s.db.Update(operation.RemoveBlob(dInfo.Height, dInfo.Cid))
```

Comment:
this is individually deleting all of the rows instead of doing it in a batch. check out the `BatchRemove*` methods on other storage objects

Reply:
[d6f8b9a](https://github.com/onflow/flow-go/pull/6277/commits/d6f8b9ab7d6f2d10a57be71b4383812b8cb99d79)